File: framework/languages/java/samples/java/DeliveryDispatch/Server/CustomerGateway/src/main/java/systems/zlink/samples/deliverydispatch/server/customergateway/handlers/CustomerStatusPushHandler.java
package systems.zlink.samples.deliverydispatch.server.customergateway.handlers;

import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.deliverydispatch.server.customergateway.CustomerActorDirectory;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup("customer-route")
public final class CustomerStatusPushHandler
    implements ZLinkRequestHandler<Messages.DeliveryStatusChangedReq, Messages.DeliveryStatusChangedRes> {
    private final CustomerActorDirectory customers;

    public CustomerStatusPushHandler(CustomerActorDirectory customers) {
        this.customers = customers;
    }

    @Override
    public CompletionStage<Messages.DeliveryStatusChangedRes> handle(
        Messages.DeliveryStatusChangedReq request,
        ZLinkRequestContext context) {
        customers.push(request);
        return CompletableFuture.completedFuture(
            new Messages.DeliveryStatusChangedRes(request.deliveryId(), request.status()));
    }
}

File: framework/languages/java/samples/kotlin/DeliveryDispatch/Server/CustomerGateway/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/customergateway/handlers/CustomerStatusPushHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.handlers

import systems.zlink.framework.channels.ZLinkRequestContext
import systems.zlink.framework.channels.ZLinkRequestHandler
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.CustomerActorDirectory
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatusChangedRes
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatusChangedReq

@ZLinkHandlerGroup("customer-route")
class CustomerStatusPushHandler(
    private val customers: CustomerActorDirectory,
) : ZLinkRequestHandler<DeliveryStatusChangedReq, DeliveryStatusChangedRes> {
    override fun handle(
        request: DeliveryStatusChangedReq,
        context: ZLinkRequestContext,
    ): DeliveryStatusChangedRes {
        customers.push(request)
        return DeliveryStatusChangedRes(request.deliveryId, request.status)
    }
}
